Refactor template options logic into `Templates::TemplateOptionService` and update template validation in `PlansController#create`

aaronskiba opened the pull request:
# Changes proposed in this PR:
[Refactor plan.org resolution into private helper](https://github.com/portagenetwork/roadmap/pull/1228/commits/005ca54eb92b3850a0efefe11de2d8b7cf931f3d)
Extracts the logic for resolving @plan.org from plan_params into a private method `resolve_plan_org`.

[Refactor template selection logic into TemplateOptionsService](https://github.com/portagenetwork/roadmap/pull/1228/commits/c3b54a06305147e755129e9205899d982b951b42)
Move the template-loading and customization logic from TemplateOptionsController#index into a new service, Templates::TemplateOptionService#available_templates.
- Makes template logic reusable for other controllers (e.g., PlansController#create).
- Removed `DEFAULT_FUNDER_ID` assignment in `TemplateOptionsController`.  After this refactor, `Rails.application.config.default_funder_id` is only needed in that controller's `build_templates_payload` method. It is now called there explicitly.

[Validate plan.template against available templates](https://github.com/portagenetwork/roadmap/pull/1228/commits/e8986f3d9b4d0e5bad9e76502369e9f245dda91b)
Update `PlansController#find_valid_template` to ensure the selected template_id belongs to the set returned by `Templates::TemplateOptionService.available_templates(org)`.

[Update and extend PlansController request specs](https://github.com/portagenetwork/roadmap/pull/1228/commits/990ed808d660a4194516eceee1fabf97bcff6b62)
Update and extend the PlansController request specs to match the new server-side `find_valid_template` behaviour. Tests now explicitly cover:
- Valid vs. invalid template/org combinations using `shared_examples`
- Default org resolution (`plan.org == current_user.org`) when no org param is supplied
- Explicit org param handling for template validation
- Ensuring `Template.default` evaluates to the seeded default template

[Update default funder check / use .find_by()](https://github.com/portagenetwork/roadmap/pull/1228/commits/55a8b3e8b56e765bf750b7630b54702ad1ed1cd6)
The following guard exists on line 13 of this file:
```ruby
return templates unless (org.present? && !org.new_record?) || (funder.present? && !funder.new_record?)
```
Rather than evaluating to nil, `Org.find(Rails.application.config.default_funder_id)` risks raising `ActiveRecord::RecordNotFound`.
`Org.find_by(id: Rails.application.config.default_funder_id)` allows us to use the guard if the funder ever does not resolve to an actual Org.

aaronskiba committed:
Refactor plan.org resolution into private helper

Extracts the logic for resolving @plan.org from plan_params into a private method `resolve_plan_org`.

aaronskiba committed:
Refactor template selection logic into TemplateOptionsService

Move the template-loading and customization logic from TemplateOptionsController#index into a new service, Templates::TemplateOptionService#available_templates.
- Makes template logic reusable for other controllers (e.g., PlansController#create).
- Removed `DEFAULT_FUNDER_ID` assignment in `TemplateOptionsController`.  After this refactor, `Rails.application.config.default_funder_id` is only needed in that controller's `build_templates_payload` method. It is now called there explicitly.

aaronskiba committed:
Validate plan.template against available templates

Update `PlansController#find_valid_template` to ensure the selected template_id belongs to the set returned by `Templates::TemplateOptionService.available_templates(org)`.

aaronskiba committed:
Use local variable and always return array in TemplateOptionsService.available_templates

- Replace the class instance variable @templates with a local templates variable to avoid shared state.
- Update top level return to return an array rather than nil
  - `TemplateOptionsController#index` and `PlansController#find_valid_template` each expect an array.

momo3404 commented on app/controllers/template_options_controller.rb:
```diff
@@ -11,56 +11,19 @@ class TemplateOptionsController < ApplicationController
     authorize Template.new, :template_options?
 
     org = org_from_params(params_in: { org_id: org_hash.to_json }) if org_hash.present?
-    funder = Org.find(DEFAULT_FUNDER_ID)
```
Doesn't seem like there's a need for `DEFAULT_FUNDER_ID` in `template_options_controller` anymore. If you want, you could just replace the one instance of it in `build_templates_payload`.

aaronskiba replied:
Cool, I like that idea and will do it.

aaronskiba replied:
I guess it makes sense to even try to add that to the initial refactor commit. I went ahead and did that (https://github.com/portagenetwork/roadmap/compare/42dafa77a5f933406e4d41146b0f69d2e921cfe9..d8e5f4830d2087f4cfc804c2586924b45d1fd694).

momo3404 commented on app/services/templates/template_options_service.rb:
```diff
@@ -0,0 +1,50 @@
+
+      templates = []
+
+      return templates unless (org.present? && !org.new_record?) || (funder.present? && !funder.new_record?)
```
I'm realizing now I never fully understood this line. I see that `new_record` returns true if  the org hasn't been saved yet. It seems like it's way to ensure that the user has picked a valid org and funder, but I just wanted to double check.

aaronskiba replied:
Yes, that sounds right; a validation check to ensure the specified org and funder are already existing in the db.

aaronskiba committed:
Update and extend `PlansController` request specs

Update and extend the PlansController request specs to match the new server-side `find_valid_template` behaviour. Tests now explicitly cover:
- Valid vs. invalid template/org combinations using `shared_examples`
- Default org resolution (`plan.org == current_user.org`) when no org param is supplied
- Explicit org param handling for template validation
- Ensuring `Template.default` evaluates to the seeded default template

momo3404 reviewed (dismissed):
Really nice refactor and clean, important tests. Nice work!

aaronskiba committed:
Update default funder check / use .find_by()

The following guard exists on line 13 of this file:
return templates unless (org.present? && !org.new_record?) || (funder.present? && !funder.new_record?)

Rather than evaluating to nil, Org.find(Rails.application.config.default_funder_id) risks raising ActiveRecord::RecordNotFound.
Org.find_by(id: Rails.application.config.default_funder_id) allows us to use the guard if the funder ever does not resolve to an actual org.

aaronskiba committed:
Update CHANGELOG.md
